File: iosApp/iosApp/Modules/CurrencyAdd/CurrencyAddWireframeFactory.swift
// Created by web3d4v on 20/06/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

// MARK: - CurrencyAddWireframeFactory

protocol CurrencyAddWireframeFactory {
    func make(
        _ parent: UIViewController?,
        context: CurrencyAddWireframeContext
    ) -> CurrencyAddWireframe
}

// MARK: - DefaultCurrencyAddWireframeFactory

final class DefaultCurrencyAddWireframeFactory {
    private let networkPickerWireframeFactory: NetworkPickerWireframeFactory
    private let qrCodeScanWireframeFactory: QRCodeScanWireframeFactory
    private let currencyStoreService: CurrencyStoreService

    init(
        networkPickerWireframeFactory: NetworkPickerWireframeFactory,
        qrCodeScanWireframeFactory: QRCodeScanWireframeFactory,
        currencyStoreService: CurrencyStoreService
    ) {
        self.networkPickerWireframeFactory = networkPickerWireframeFactory
        self.qrCodeScanWireframeFactory = qrCodeScanWireframeFactory
        self.currencyStoreService = currencyStoreService
    }
}

extension DefaultCurrencyAddWireframeFactory: CurrencyAddWireframeFactory {

    func make(
        _ parent: UIViewController?,
        context: CurrencyAddWireframeContext
    ) -> CurrencyAddWireframe {
        DefaultCurrencyAddWireframe(
            parent,
            context: context,
            networkPickerWireframeFactory: networkPickerWireframeFactory,
            qrCodeScanWireframeFactory: qrCodeScanWireframeFactory,
            currencyStoreService: currencyStoreService
        )
    }
}

// MARK: - Assembler

final class CurrencyAddWireframeFactoryAssembler: AssemblerComponent {
    
    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> CurrencyAddWireframeFactory in
            DefaultCurrencyAddWireframeFactory(
                networkPickerWireframeFactory: resolver.resolve(),
                qrCodeScanWireframeFactory: resolver.resolve(),
                currencyStoreService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/currencyadd/CurrencyAddWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.currencyadd

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3walletcore.modules.currencyAdd.CurrencyAddWireframe
import com.sonsofcrypto.web3walletcore.modules.currencyAdd.CurrencyAddWireframeContext
import smartadapter.internal.extension.name


interface CurrencyAddWireframeFactory {
    fun make(parent: Fragment?, context: CurrencyAddWireframeContext): CurrencyAddWireframe
}

class DefaultCurrencyAddWireframeFactory(
    private val currencyStoreService: CurrencyStoreService,
): CurrencyAddWireframeFactory {

    override fun make(
        parent: Fragment?, context: CurrencyAddWireframeContext
    ): CurrencyAddWireframe = DefaultCurrencyAddWireframe(
        parent?.let { WeakRef(it) },
        context,
        currencyStoreService,
    )
}

class CurrencyAddWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(CurrencyAddWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultCurrencyAddWireframeFactory(
                it.resolve(CurrencyStoreService::class.name),
            )
        }
    }
}